File: scripts/doc2query_roda.ts
#!/usr/bin/env bun
/**
 * Roda o doc2query nos 33 lotes via `opencode run`, valida cada saída e retoma
 * de onde parou.
 *
 *   bun run scripts/doc2query_roda.ts                   # todos os lotes pendentes
 *   bun run scripts/doc2query_roda.ts --lote 00_amostra # só um
 *   bun run scripts/doc2query_roda.ts --modelo anthropic/claude-sonnet-4-5
 *   bun run scripts/doc2query_roda.ts --revalidar       # só confere o que já existe
 *
 * É retomável de propósito: são 33 chamadas de LLM, e uma queda no lote 20 não
 * pode custar os 19 anteriores. Lote com saída válida é pulado.
 */
import { readFileSync, writeFileSync, existsSync, readdirSync, unlinkSync } from "node:fs";

const DIR = "tasks/doc2query";
const PROMPT = "scripts/prompts/doc2query.md";
const arg = (n: string) => { const i = Bun.argv.indexOf(n); return i > 0 ? Bun.argv[i + 1] : undefined; };
const MODELO = arg("--modelo");
const SO_LOTE = arg("--lote");
const REVALIDAR = Bun.argv.includes("--revalidar");
const PERGUNTAS_ESPERADAS = 8;

interface Saida { id: string; perguntas: string[]; incerta?: boolean }

/**
 * Valida a saída de um lote. Rejeitar cedo importa: uma saída ruim não dá erro,
 * ela envenena a busca em silêncio — que é exatamente o problema que estamos
 * consertando.
 */
function validar(loteFile: string, saidaFile: string): { ok: boolean; erros: string[]; n: number } {
  const erros: string[] = [];
  if (!existsSync(saidaFile)) return { ok: false, erros: ["saída não existe"], n: 0 };

  const entrada: any[] = readFileSync(loteFile, "utf-8").trim().split("\n").map((l) => JSON.parse(l));
  const esperados = new Map<string, string[]>(
    entrada.map((e: any) => [e.id as string, (e.colunas as string[]).map((c) => c.toLowerCase())]));

  let saidas: Saida[];
  try {
    saidas = readFileSync(saidaFile, "utf-8").trim().split("\n")
      .filter((l) => l.trim() && !l.trim().startsWith("```"))
      .map((l) => JSON.parse(l));
  } catch (e) {
    return { ok: false, erros: [`JSON inválido: ${e}`], n: 0 };
  }

  const vistos = new Set<string>();
  let total = 0;
  for (const s of saidas) {
    if (!esperados.has(s.id)) { erros.push(`id fora do lote: ${s.id}`); continue; }
    if (vistos.has(s.id)) erros.push(`id repetido: ${s.id}`);
    vistos.add(s.id);

    if (!Array.isArray(s.perguntas) || s.perguntas.length === 0) { erros.push(`${s.id}: sem perguntas`); continue; }
    if (!s.incerta && s.perguntas.length !== PERGUNTAS_ESPERADAS) {
      erros.push(`${s.id}: ${s.perguntas.length} perguntas, esperado ${PERGUNTAS_ESPERADAS}`);
    }
    total += s.perguntas.length;

    const cols = esperados.get(s.id)!;
    for (const q of s.perguntas) {
      if (typeof q !== "string" || q.length < 8) { erros.push(`${s.id}: pergunta curta demais: "${q}"`); continue; }
      if (q.split(/\s+/).length > 20) erros.push(`${s.id}: pergunta longa demais: "${q.slice(0, 50)}…"`);
      // A regra 1 do prompt, verificada: ecoar nome de coluna anula o exercício,
      // porque o índice já tem os nomes e é justamente isso que não funciona.
      const eco = cols.find((c) => c.includes("_") && q.toLowerCase().includes(c));
      if (eco) erros.push(`${s.id}: ecoa nome de coluna "${eco}" em "${q.slice(0, 46)}…"`);
    }
  }

  for (const id of esperados.keys()) if (!vistos.has(id)) erros.push(`faltou: ${id}`);
  return { ok: erros.length === 0, erros, n: total };
}

async function rodarLote(nome: string): Promise<boolean> {
  const lote = `${DIR}/lote_${nome}.jsonl`;
  const saida = `${DIR}/saida_${nome}.jsonl`;

  const msg =
    `Leia ${PROMPT} e siga-o à risca para todas as tabelas de ${lote}. ` +
    `Escreva o resultado em ${saida}: um JSON por linha, na mesma ordem da entrada, ` +
    `sem cercas de markdown e sem nenhum texto fora do JSONL. ` +
    `Não rode nenhum comando além de ler a entrada e escrever a saída.`;

  const args = ["run", "--auto", ...(MODELO ? ["--model", MODELO] : []), msg];
  const proc = Bun.spawn(["opencode", ...args], { stdout: "pipe", stderr: "pipe" });
  const [out, err, code] = await Promise.all([
    new Response(proc.stdout).text(), new Response(proc.stderr).text(), proc.exited,
  ]);
  if (code !== 0) { console.log(`  opencode saiu com ${code}: ${err.trim().slice(0, 300)}`); return false; }
  if (!existsSync(saida)) {
    // O agente rodou mas nao gravou. Acontece com modelo que "responde" em vez
    // de usar a ferramenta de escrita — guarda o log pra isso nao ser mudo.
    writeFileSync(`${DIR}/erro_${nome}.log`, out.slice(-4000));
    console.log(`  o agente nao gravou ${saida} — log em ${DIR}/erro_${nome}.log`);
    return false;
  }

  const v = validar(lote, saida);
  if (!v.ok) {
    console.log(`  INVÁLIDO (${v.erros.length} problema(s)):`);
    for (const e of v.erros.slice(0, 6)) console.log(`    ${e}`);
    if (v.erros.length > 6) console.log(`    … e mais ${v.erros.length - 6}`);
    // apaga: saída meio-boa é pior que ausente, porque parece pronta
    if (existsSync(saida)) unlinkSync(saida);
    return false;
  }
  console.log(`  ok — ${v.n} perguntas`);
  return true;
}

const lotes = SO_LOTE ? [SO_LOTE]
  : readdirSync(DIR).filter((f) => f.startsWith("lote_")).sort()
      .map((f) => f.replace("lote_", "").replace(".jsonl", ""));

if (REVALIDAR) {
  let bons = 0, total = 0;
  for (const n of lotes) {
    const v = validar(`${DIR}/lote_${n}.jsonl`, `${DIR}/saida_${n}.jsonl`);
    if (v.ok) { bons++; total += v.n; }
    else if (existsSync(`${DIR}/saida_${n}.jsonl`)) console.log(`lote ${n}: ${v.erros[0]}`);
  }
  console.log(`\n${bons}/${lotes.length} lotes válidos, ${total} perguntas`);
  process.exit(0);
}

let feitos = 0, falhos = 0;
for (const n of lotes) {
  if (validar(`${DIR}/lote_${n}.jsonl`, `${DIR}/saida_${n}.jsonl`).ok) { feitos++; continue; }
  console.log(`lote ${n} …`);
  if (await rodarLote(n)) feitos++; else falhos++;
}
console.log(`\n${feitos}/${lotes.length} lotes prontos` + (falhos ? `, ${falhos} falharam — rode de novo, ele retoma` : ""));

File: scripts/doc2query_roda.py
#!/usr/bin/env python3
"""Resumable doc2query batch runner — Python port of ask-web's doc2query_roda.ts,
adapted to call `opencode run --model opencode/hy3-free` (OpenCode Zen free tier;
`opencode-go/*` is out of balance, `opencode/hy3-free` is not).

    python3 doc2query_roda.py              # all pending lotes
    python3 doc2query_roda.py --lote 05    # just one

Retomable de propósito: uma queda no lote 20 nao pode custar os 19 anteriores.
Lote com saida ja valida e pulado.
"""
import argparse
import json
import subprocess
import sys
import time
from pathlib import Path

DIR = Path("tasks/doc2query")
PROMPT = Path("scripts/prompts/doc2query.md")
MODEL = "opencode/hy3-free"
OPENCODE_BIN = "opencode"
PERGUNTAS_ESPERADAS = 8
ESTRUTURAIS = {"nome_coluna", "id_tabela", "cobertura_temporal"}


def validar(lote_file: Path, saida_file: Path):
    entrada = [json.loads(l) for l in lote_file.read_text(encoding="utf-8").splitlines() if l.strip()]
    esperados = {e["id"]: [c.lower() for c in e["colunas"]] for e in entrada}
    erros = []
    if not saida_file.exists():
        return {"ok": False, "erros": ["saida nao existe"], "n": 0}
    try:
        linhas = [l for l in saida_file.read_text(encoding="utf-8").strip().split("\n")
                  if l.strip() and not l.strip().startswith("```")]
        saidas = [json.loads(l) for l in linhas]
    except Exception as e:
        return {"ok": False, "erros": [f"JSON invalido: {e}"], "n": 0}

    vistos = set()
    total = 0
    for s in saidas:
        if s["id"] not in esperados:
            erros.append(f"id fora do lote: {s['id']}")
            continue
        if s["id"] in vistos:
            erros.append(f"id repetido: {s['id']}")
        vistos.add(s["id"])
        cols = esperados[s["id"]]
        perguntas = s.get("perguntas", [])
        if not perguntas:
            erros.append(f"{s['id']}: sem perguntas")
            continue
        if not s.get("incerta") and len(perguntas) != PERGUNTAS_ESPERADAS:
            erros.append(f"{s['id']}: {len(perguntas)} perguntas, esperado {PERGUNTAS_ESPERADAS}")
        eh_dicionario = s["id"].endswith(".dicionario")
        for q in perguntas:
            if not isinstance(q, str) or len(q) < 8:
                erros.append(f"{s['id']}: pergunta curta demais: {q!r}")
                continue
            if len(q.split()) > 20:
                erros.append(f"{s['id']}: pergunta longa demais: {q[:50]!r}")
            ql = q.lower()
            for c in cols:
                if "_" in c and c in ql and not (eh_dicionario and c in ESTRUTURAIS):
                    erros.append(f"{s['id']}: ecoa coluna '{c}' em {q[:46]!r}")
                    break
        total += len(perguntas)
    for id_ in esperados:
        if id_ not in vistos:
            erros.append(f"faltou: {id_}")
    return {"ok": len(erros) == 0, "erros": erros, "n": total}


def rodar_lote(nome: str) -> bool:
    lote = DIR / f"lote_{nome}.jsonl"
    saida = DIR / f"saida_{nome}.jsonl"

    msg = (
        f"Leia {PROMPT} e siga-o à risca para todas as tabelas de {lote}. "
        f"Escreva o resultado em {saida}: um JSON por linha, na mesma ordem da entrada, "
        f"sem cercas de markdown e sem nenhum texto fora do JSONL. "
        f"Não rode nenhum comando além de ler a entrada e escrever a saída."
    )
    cmd = [OPENCODE_BIN, "run", "--auto", "--model", MODEL, msg]
    try:
        proc = subprocess.run(cmd, capture_output=True, text=True, timeout=600)
    except subprocess.TimeoutExpired:
        print("  timeout apos 600s")
        return False

    if proc.returncode != 0:
        print(f"  opencode saiu com {proc.returncode}: {proc.stderr.strip()[:300]}")
        return False
    if not saida.exists():
        errlog = DIR / f"erro_{nome}.log"
        errlog.write_text(proc.stdout[-4000:], encoding="utf-8")
        print(f"  o agente nao gravou {saida} — log em {errlog}")
        return False

    v = validar(lote, saida)
    if not v["ok"]:
        print(f"  INVALIDO ({len(v['erros'])} problema(s)):")
        for e in v["erros"][:6]:
            print(f"    {e}")
        if len(v["erros"]) > 6:
            print(f"    ... e mais {len(v['erros']) - 6}")
        saida.unlink(missing_ok=True)
        return False
    print(f"  ok — {v['n']} perguntas")
    return True


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--lote")
    ap.add_argument("--max-retries", type=int, default=2)
    args = ap.parse_args()

    if args.lote:
        lotes = [args.lote]
    else:
        lotes = sorted(
            f.stem.replace("lote_", "") for f in DIR.glob("lote_*.jsonl")
        )

    feitos, falhos = 0, 0
    for n in lotes:
        lote_file, saida_file = DIR / f"lote_{n}.jsonl", DIR / f"saida_{n}.jsonl"
        if validar(lote_file, saida_file)["ok"]:
            feitos += 1
            print(f"lote {n} — ja pronto, pulando")
            continue
        print(f"lote {n} ...", flush=True)
        ok = False
        for attempt in range(args.max_retries):
            if rodar_lote(n):
                ok = True
                break
            print(f"  retry {attempt + 1}/{args.max_retries}")
            time.sleep(5)
        if ok:
            feitos += 1
        else:
            falhos += 1
        sys.stdout.flush()

    print(f"\n{feitos}/{len(lotes)} lotes prontos" +
          (f", {falhos} falharam apos retries" if falhos else ""))


if __name__ == "__main__":
    main()
